design/open/graveler_stateful_entities.md

```diff
@@ -0,0 +1,61 @@
+# Graveler Stateful Entities
+
+## Abstract
+
+Transitioning from PostgresSQL based metadata management DB to [generic KV Store design](https://github.com/treeverse/lakeFS/blob/master/design/accepted/metadata_kv/index.md) 
+introduced challenges in regard to atomicity and error handling.
+Previously, we relied on the transactional property of PSQL, to perform all-or-nothing operations to perform flows which requires 
+the modification of multiple entities atomically. Together with the built-in lock mechanism we were able to perform atomic, and concurrently safe operations.
+In the KV design, we rely on a very narrow interface, which does not support transactions. Additionally, the KV design does not rely on locks for operations.
+This complicates atomicity, error handling and rollback operations.
+This design can also pave to way for future support of [long-running hooks](https://github.com/treeverse/lakeFS/issues/3020), which require entities to be in an initial in-accessible/operable state,
+until the pre-action hooks are completed, and in case of failure deleted without implications on the end user.
+Proceeding the [stateful design for repositories](https://github.com/treeverse/lakeFS/pull/3661) , this document suggests adding states for all (most) of the graveler entities to 
+support both goals.
+
+## Supported entities
+
+All entities which currently support pre/post actions on creation and deletion must support states. These include:
+* Branches
+* Commits
+* Tags
+
+All entities which require support of transactional / concurrent / rollback operations:
+* Repositories
+* Branches
+
+## Supported states
```

Comment:
State is an application level feature which is bound to the usages of the specific entity. Since every entity has a different usage patterns, I think that trying to align all entities to use the same list of states is as hard as writing the same code to handle all entities.